### PR TITLE
12/errors: basic structure

### PR DESCRIPTION
**Pull request by grrtrr**
* Talk (~14 minutes)
 * 2 exercises
 * additional resources at end

### CODE REVIEW

File: 12-errors.slide [hunk 1]
```diff
@@ -1,29 +1,210 @@
...
+- _integrate_better_in_normal_control_flow_
+- allow/require user to define policy
+
+ "Simplicity is prerequisite for reliability." -- E.W. Dijkstra
+
+* 2. Errors In Go
+
+* Proverb 1: Errors are  values
+
+Dedicated valuev/ariable:
```
Comment (nathany) [inline 12-errors.slide]: value/variable

File: 12-errors/exit_and_log.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+// fatalist indicates whether an error is severe // HL
+type fatalist interface {
+	IsFatal() bool // HL
+}
+
+func main() {
+	err := KickThunderbolt()
+	if err != nil {
+		if f, ok := err.(fatalist); ok && f.IsFatal() { // HL
```
Comment (nathany) [inline 12-errors/exit_and_log.go]: be prepared to explain type assertions, as I don't know that they will be covered before this

File: 12-errors/retry.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	globalVariable++
+	if globalVariable > 3 {
+		return nil
+	}
+	return SysIntError(globalVariable)
+}
+func main() {
+	for attempt := 1; attempt <= 5; attempt++ {
+		err := RenderGraphics() // HL
+		if err == nil {         // HL
```
Comment (nathany) [inline 12-errors/retry.go]: I'm not sure if I would do this any differently, but it is interesting in that the happy path is indented here.

File: 12-errors/scanner.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	var scanner = bufio.NewScanner(strings.NewReader(input))
+
+	scanner.Split(bufio.ScanWords)
+
+	for scanner.Scan() { // HL
+		count++
+		fmt.Printf("%d: %s\n", count, scanner.Text())
+	} // HL
+
+	if err := scanner.Err(); err != nil { // Error tested at end // HL
```
Comment (nathany) [inline 12-errors/scanner.go]: 👍

File: 12-errors.slide [hunk 5]
```diff
@@ -1,29 +1,222 @@
...
 
-error interface
+   func (c ConnectionError) Timeout()   bool // true for syscall.ETIMEDOUT
+   func (c ConnectionError) Temporary() bool // true for syscall.EBUSY, syscall.EINPROGRESS
+   func (c ConnectionError) IsFatal()   bool // true for syscall.EHOSTDOWN, syscall.ENOTRECOVERABLE
 
-errors.New, fmt.Errorf
-var Err* sentinel errors
-custom ThatError struct and Error()
+.link https://play.golang.org/p/00bJqeBq1a Playground template
```
Comment (nathany) [inline 12-errors.slide]: maybe include the link as text instead/also so people can type it in

File: 12-errors/exit_and_log.go (hunk 2 above)
Comment (grrtrr) [inline 12-errors/exit_and_log.go]: Wanted to make this a separate slide, but ran out of space; not intending too much depth here.

**Comment by grrtrr** [comment]
Thank you for the review, will address immediate comments now; anything further would be in a separate PR.

File: 12-errors/exit_and_log.go (hunk 2 above)
Comment (nathany) [inline 12-errors/exit_and_log.go]: probably no need to add a slide, just so long as you're ready for the question

**Comment by nathany** [comment]
Thanks Gerrit. Please hit merge whenever you're ready.

**Comment by grrtrr** [comment]
Thanks again ... just did

**Comment by nathany** [comment]
btw, you don't need a fork of this repo because you have full access to it. you can just make a branch and push it before opening a pull request.

**Pull request merged**